janusgraph-core/src/main/java/org/janusgraph/diskstorage/log/kcvs/KCVSLog.java

```diff
@@ -441,7 +455,7 @@ private KCVSMessage parseMessage(Entry msg) {
      * @param persistor
      * @return
      */
-    private Future<Message> add(StaticBuffer content, int partitionId, ExternalPersistor persistor) {
+    private synchronized Future<Message> add(StaticBuffer content, int partitionId, ExternalPersistor persistor) {
```

Comment:
Why synchronized? I suspect this might bring great overhead when WAL is enabled (i.e. SYSTEM_LOG_TRANSACTIONS == True).

Reply:
Great observation! I noticed that all the other methods which access volatile variable `isOpen` are synchronized. This methods checks that as well, so, I thought it makes sense to make it synchronized. I need to double check if we need to access isOpen here or if it's not critical to execute this method after closure

Reply:
If `isOpen` needs to be synchronized, we probably should synchronize `verifyOpen` method only. This `add` method can be called by multiple threads concurrently so synchronizing the whole method could bring great overhead.

Reply:
I don't think it makes much sense synchronizing verifyOpen method. The main idea of synchronizing is to acquire the lock, check condition, process logic, release the lock.
If we are synchronizing verifyOpen method the flow will be:
acquire the lock, check condition, release lock, process logic. Which means logic processing isn't protected.
Nevertheless, if this method can be executed in parallel then we definitely don't need to synchronize is with simple lock. I think I have an idea for this issue. I will change the logic here

Reply:
It looks like it's not critical to have synchronization here. We will simply throw `JanusGraphException` instead of `ResourceUnavailableException` in case of concurrency happens here.
I.e. in case we check that `isOpen == true` we enter `add` method, after that another thread triggers `close()` method and there is a chance of storage being closed during that time (the same may happen even when we don't trigger `close` method. For example, storage backend connection is lost). In such a situation there is a chance of trying to send a message when we don't have a backend connection anymore which will result in `JanusGraphException`. Even if this threads retries to executor `add` method right after receiving `JanusGraphException` (for any reason) we will send `ResourceUnavailableException` on the second execution because `isOpen` variable is `volatile`. Thus, I think it's OK to not having any synchronization mechanism here (i.e. leave it as it is).
I will remove `synchronize` from this method but if anyone thinks we need some synchronization here, I'm happy to discuss another better synchronization solution for this use case (i.e. we could simple have atomic integer which counts amount of current enters to this method. If so, in `close()` method after changing `isOpen` to `false` we could simple await until that variable is equal to `0`).

Reply:
> I don't think it makes much sense synchronizing verifyOpen method

You are right, I somehow thought you wanted to control concurrent access to `if (!isOpen)` block in `verifyOpen` method. Now it's clear to me. I also agree that it might be a better idea to leave it as it is, considering the tradeoff between performance and consistency.